scripts/experiment_1/functions/vae.py

```diff
@@ -0,0 +1,322 @@
+        latent_dim, kernel_initializer='glorot_uniform')(rnaseq_input)
+    z_log_var_dense_batchnorm = BatchNormalization()(z_log_var_dense_linear)
+    z_log_var_encoded = Activation('relu')(z_log_var_dense_batchnorm)
+
+    # Customized layer
+    # Returns the encoded and randomly sampled z vector
+    # Takes two keras layers as input to the custom sampling function layer with a
+    # latent_dim` output
+    #
+    # sampling():
+    # randomly sample similar points z from the latent normal distribution that is assumed to generate the data,
+    # via z = z_mean + exp(z_log_sigma) * epsilon, where epsilon is a random normal tensor
+    # z ~ Q(z|X)
+    # Note: there is a trick to reparameterize to standard normal distribution so that the space is differentiable and
+    # therefore gradient descent can be used
+    #
+    # Returns the encoded and randomly sampled z vector
+    # Takes two keras layers as input to the custom sampling function layer with a
+    # latent_dim` output
+    z = Lambda(sampling_maker(epsilon_std),
+               output_shape=(latent_dim, ))([z_mean_encoded, z_log_var_encoded])
+
+    # DECODER
+
+    # The decoding layer is much simpler with a single layer glorot uniform
+    # initialized and sigmoid activation
+    # Reconstruct P(X|z)
+    decoder_model = Sequential()
+    decoder_model.add(
+        Dense(intermediate_dim, activation='relu', input_dim=latent_dim))
+    decoder_model.add(Dense(original_dim, activation='sigmoid'))
+    rnaseq_reconstruct = decoder_model(z)
+
+    # CONNECTIONS
+    # fully-connected network
+    adam = optimizers.Adam(lr=learning_rate)
+    vae_layer = CustomVariationalLayer(original_dim, z_log_var_encoded, z_mean_encoded, beta)([
+        rnaseq_input, rnaseq_reconstruct])
+    vae = Model(rnaseq_input, vae_layer)
+    vae.compile(optimizer=adam, loss=None, loss_weights=[beta])
+
+    # Training
+
+    # fit Model
+    # hist: record of the training loss at each epoch
+    hist = vae.fit(
+        np.array(rnaseq_train_df),
+        shuffle=True,
+        epochs=epochs,
+        batch_size=batch_size,
+        validation_data=(np.array(rnaseq_test_df), None),
+        callbacks=[WarmUpCallback(beta, kappa)])
+
+    # Use trained model to make predictions
+    encoder = Model(rnaseq_input, z_mean_encoded)
+
+    encoded_rnaseq_df = encoder.predict_on_batch(rnaseq)
+    encoded_rnaseq_df = pd.DataFrame(encoded_rnaseq_df, index=rnaseq.index)
+
+    encoded_rnaseq_df.columns.name = 'sample_id'
+    encoded_rnaseq_df.columns = encoded_rnaseq_df.columns + 1
+
+    # Visualize training performance
+    history_df = pd.DataFrame(hist.history)
+    ax = history_df.plot()
+    ax.set_xlabel('Epochs')
+    ax.set_ylabel('VAE Loss')
+    fig = ax.get_figure()
+    fig.savefig(hist_plot_file)
+
+    del ax, fig
+
+    # Output
+
+    # Save training performance
+    history_df = pd.DataFrame(hist.history)
+    history_df = history_df.assign(learning_rate=learning_rate)
```

Comment:
Since they are not doing any new operations, they can just be added rather than using the assign function